File: 2023/11/11.py

#shortest path between every pair of galaxies
with open('11test.txt', 'r') as f:
    lines = f.read().split('\n')

def find_galaxies(lines):
    galaxies = [(i, j) for i, line in enumerate(lines) for j, char in enumerate(line) if char == '#']
    space = [(i, j) for i, line in enumerate(lines) for j, char in enumerate(line) if char != '#']

    return galaxies, space

def find_empty_spaces(lines):
    empty_rows = [i for i, line in enumerate(lines) if all(char != '#' for char in line)]
    
    # Transpose the lines to find empty columns
    transposed_lines = list(map(list, zip(*lines)))
    empty_cols = [j for j, col in enumerate(transposed_lines) if all(char != '#' for char in col)]

    return empty_rows, empty_cols



galaxies, space = find_galaxies(lines)
x, y = find_empty_spaces(lines)


def get_galaxy_pairs(galaxies):
    galaxy_pairs = []

    for i, (galaxy_row, galaxy_col) in enumerate(galaxies):
        for (other_galaxy_row, other_galaxy_col) in galaxies[:i]:
            galaxy_pairs.append(((galaxy_row, galaxy_col), (other_galaxy_row, other_galaxy_col)))

    return galaxy_pairs

def shortest_path_between_pairs(galaxy_pairs):
    
    shortest_path = 0

    for (galaxy_row, galaxy_col), (other_galaxy_row, other_galaxy_col) in galaxy_pairs:
        for row in range(min(other_galaxy_row, galaxy_row), max(other_galaxy_row, galaxy_row)):
            shortest_path += 2 if row in x else 1

        for col in range(min(other_galaxy_col, galaxy_col), max(other_galaxy_col, galaxy_col)):
            shortest_path += 2 if col in y else 1

    return shortest_path

print(get_galaxy_pairs(galaxies))
print(len(get_galaxy_pairs(galaxies)))
print(shortest_path_between_pairs(get_galaxy_pairs(galaxies)))

File: 2023/11/11.scala
import scala.io.Source


object Galaxies {

    def findGalaxies(lines: Array[String]) : (Array[Int, Int]) = {

            val galaxies = for {
                (line, i) <- lines.zipWithIndex
                (char, j) <- line.zipWithIndex
                if char == '#'
            } yield (i, j)

            val space = for {
                (line, i) <- lines.zipWithIndex
                (char, j) <- line.zipWithIndex
                if char == '.'
            } yield (i, j)
            
            (galaxies, space)
        }

    def findEmptySpaces(lines: Array[String]): (Array[Int], Array[Int]) = {

        val emptyRows = lines.zipWithIndex.collect{
            case (line, i) if line.forAll(_ != '#') => i
        }

        val transposedLines = lines.transpose
        val emptyCols = transposedLines.zipWithIndex.collect{
            case (col, i) if col.forAll(_ != '#') => i
        }

        (emptyRows, emptyCols)
    }

    val (galaxies, space) = findGalaxies(lines)
    val (x, y) = findEmptySpaces(lines)


    def getGalaxyPairs(galaxies: Array[(Int, Int)]): Array[((Int, Int), (Int, Int))] = {
        val pairs = for {
            (galaxy, i) <- galaxies.zipWithIndex
            (otherGalaxy, j) <- galaxies.take(i)
            if i < j
        } yield (galaxy, otherGalaxy)
    }

    // Shortest path between pairs (functional approach)
    def shortestPathBetweenPairsFunctional(galaxyPairs: Array[((Int, Int), (Int, Int))], x: Array[Int], y: Array[Int]): Int = {
         
        // Calculate path between two galaxies (one pair) - then sum
        def calculatePath(galaxyPair: ((Int, Int), (Int, Int))): Int = {
            val ((galaxyRow, galaxyCol), (otherGalaxyRow, otherGalaxyCol)) = galaxyPair

            def calculateDimensionPath(min: Int, max: Int, coordinates: Array[Int]): Int =
            (min until max).map(coord => if (coordinates.contains(coord)) 2 else 1).sum

            calculateDimensionPath(math.min(otherGalaxyRow, galaxyRow), math.max(otherGalaxyRow, galaxyRow), x) +
            calculateDimensionPath(math.min(otherGalaxyCol, galaxyCol), math.max(otherGalaxyCol, galaxyCol), y)
        }

        galaxyPairs.map(calculatePath).sum
    }

    // Usage
    val result = shortestPathBetweenPairsFunctional(galaxyPairs, x, y)
    println(result)


}
